### DISCUSSION TITLE
Step-Time Training Plugin: Measure GPU Step Latency Without Recompiling vLLM

### DISCUSSION THREAD

**Discussion by sriumcp** [inference-sim/inference-sim#1230] [Ideas]
## What This Is

This document shows how to collect per-step GPU execution timing data from a running vLLM instance — across any batch shape, with tensor parallelism (TP) or expert parallelism (EP) enabled — **without modifying or recompiling vLLM**.

The intended use case is training latency estimation models (e.g. roofline-based physics models) that predict step time from batch shape and hardware configuration. The same technique is useful for any project that needs ground-truth GPU timing per forward pass.

---

## Background: What Is a "Step"?

Every time vLLM schedules a batch of requests, it calls `GPUModelRunner._model_forward()` once. That single call — one transformer forward pass over the current batch — is a **step**. Its duration is the "step time."

Step time depends on:

| Dimension | Why it matters |
|---|---|
| Total prefill tokens | Quadratic attention cost during prompt processing |
| Number of decode requests | Each generates exactly one new token per step |
| Sum of context lengths across decode requests | KV cache read bandwidth scales linearly with past tokens |
| Batch size (total requests) | Per-request scheduling overhead |
| Hardware (GPU type, tensor-parallel degree) | Sets the compute and bandwidth ceilings |

A latency model trained on measurements across these dimensions can accurately predict step time for capacity planning without requiring live GPUs.

---

## Why Not Edit the Source?

Compiling vLLM requires CUDA toolchain setup, can take 20–40 minutes, and needs to be redone on every cluster node after updates. The plugin approach avoids all of that:

- The plugin is a **pure-Python package** (~30 lines)
- It installs with `pip install -e .` in seconds
- It uses vLLM's built-in plugin system, which loads the plugin in every worker process automatically
- Removing the plugin is `pip uninstall blis-timing-plugin`

---

## How vLLM's Plugin System Works

vLLM loads plugins via Python's standard [entry points](https://packaging.python.org/en/latest/specifications/entry-points/) mechanism. Any installed package that declares an entry point under the group `vllm.general_plugins` will have its registered function called automatically in **every worker process** (including all TP/EP GPU workers) during worker startup — before any model is loaded or any request is processed.

This is the exact hook we need: it runs inside each worker process after it has been spawned, so the patch we apply to the model runner class is the one that actually runs inference.

The loading happens at `vllm/v1/worker/worker_base.py`, in `WorkerProcessEntrypoint.__init__()`:

```python
from vllm.plugins import load_general_plugins
load_general_plugins()   # runs every registered vllm.general_plugins entry point
```

This is called before `init_device()`, before `load_model()`, and well before any `execute_model()` call.

---

## The Timing Hook

`GPUModelRunner._model_forward()` is the method that calls `self.model(...)` — the actual transformer forward pass. Its docstring says:

> *"This method can be overridden by subclasses for model execution. Motivation: We can inspect only this method versus the whole execute_model, which has additional logic."*

We monkey-patch it to wrap the call with CUDA events. CUDA events (not `time.perf_counter`) are used because they measure **GPU-side execution time** rather than CPU wall time, which includes unrelated kernel launch latency from prior steps.

---

### Step 1: Create the plugin package

Create a directory anywhere on the cluster (does not need to be inside the vLLM tree):

```
blis-timing-plugin/
├── pyproject.toml
└── blis_timing/
    └── __init__.py
```

### Step 2: Write `pyproject.toml`

```toml
[build-system]
requires = ["setuptools>=61"]
build-backend = "setuptools.backends.legacy:build"

[project]
name = "blis-timing-plugin"
version = "0.1.0"
# No dependencies — uses only packages already present in vLLM's environment.

[project.entry-points."vllm.general_plugins"]
blis_timing = "blis_timing:register"
```

The entry point declaration says: when vLLM loads general plugins, call `blis_timing.register()`.

### Step 3: Write `blis_timing/__init__.py`

```python
"""
vLLM general plugin: records per-step GPU forward-pass timing.

Activated automatically in every vLLM worker process.
Output: one JSON line per step written to $BLIS_TIMING_OUT (default /tmp/blis_timing.jsonl).

Each record:
  num_tokens     – total tokens in the forward pass (prefill + decode)
  num_reqs       – number of requests in the batch
  sum_ctx_tokens – sum of already-computed tokens across all decode requests
                   (i.e. Σ context_length for decode reqs; 0 for pure-prefill steps)
  step_time_us   – GPU execution time in microseconds (CUDA event measurement)
"""

import json
import os

import torch

_OUT_FILE = os.environ.get("BLIS_TIMING_OUT", "/tmp/blis_timing.jsonl")
_original_model_forward = None


def _timed_model_forward(
    self,
    input_ids,
    positions,
    intermediate_tensors=None,
    inputs_embeds=None,
    **model_kwargs,
):
    # is_global_first_rank() returns True on rank 0, False on all other TP/EP ranks.
    # It safely returns True when torch.distributed is not yet initialized (single-GPU).
    # We only record timing on rank 0 because:
    #   - NCCL All-Reduce (TP) blocks rank 0 until ALL TP ranks complete their shard.
    #   - NCCL All-to-All (EP) blocks rank 0 until ALL expert ranks return results.
    # So rank 0's elapsed time naturally includes TP/EP communication overhead.
    from vllm.distributed.parallel_state import is_global_first_rank

    if not is_global_first_rank():
        return _original_model_forward(
            self, input_ids, positions, intermediate_tensors, inputs_embeds,
            **model_kwargs
        )

    # Capture batch shape BEFORE the forward pass.
    # input_batch is already updated by _update_states() before _model_forward() is called.
    batch = self.input_batch
    num_reqs = batch.num_reqs
    sum_ctx = int(batch.num_computed_tokens_cpu[:num_reqs].sum())

    # CUDA events measure GPU-side execution time precisely.
    # They bracket the actual GPU kernels, not CPU scheduling overhead.
    start_event = torch.cuda.Event(enable_timing=True)
    end_event = torch.cuda.Event(enable_timing=True)

    start_event.record()  # enqueued into the default CUDA stream
    result = _original_model_forward(
        self, input_ids, positions, intermediate_tensors, inputs_embeds,
        **model_kwargs
    )
    end_event.record()    # enqueued after all forward kernels (including TP All-Reduce)

    # Block CPU until all CUDA work up to end_event is complete.
    # For TP: the last All-Reduce inside the forward pass must finish on all ranks
    # before end_event can fire on rank 0, so this synchronize() implicitly
    # waits for the entire TP group. No separate barrier needed.
    torch.cuda.synchronize()

    step_time_us = start_event.elapsed_time(end_event) * 1000.0  # ms → µs

    record = {
        "num_tokens": int(input_ids.shape[0]) if input_ids is not None else 0,
        "num_reqs": num_reqs,
        "sum_ctx_tokens": sum_ctx,
        "step_time_us": step_time_us,
    }

    with open(_OUT_FILE, "a") as f:
        f.write(json.dumps(record) + "\n")

    return result


def register():
    """
    Called by vLLM's plugin loader inside every worker process before model load.
    Patches GPUModelRunner._model_forward with a timing wrapper.

    The idempotent guard prevents double-patching when the function is called
    more than once in the same process (can happen with fork-based workers that
    inherit an already-patched class from the parent).
    """
    global _original_model_forward

    from vllm.v1.worker.gpu_model_runner import GPUModelRunner

    if GPUModelRunner._model_forward is _timed_model_forward:
        return  # already patched in this process — do nothing

    _original_model_forward = GPUModelRunner._model_forward
    GPUModelRunner._model_forward = _timed_model_forward
```

### Step 4: Install the plugin

From any directory on the cluster node (in the same Python environment vLLM uses):

```bash
pip install -e /path/to/blis-timing-plugin/
```

That is all. No vLLM recompilation. No restart of any service. The plugin is picked up the next time a vLLM worker process starts.

To verify installation:

```bash
python -c "
from importlib.metadata import entry_points
eps = entry_points(group='vllm.general_plugins')
print([e.name for e in eps])
# Should print: ['blis_timing']
"
```

To remove:

```bash
pip uninstall blis-timing-plugin
```

---

### Minimum working example

```python
import json
import os
from vllm import LLM, SamplingParams

# Point output file before creating LLM (workers inherit the env var).
os.environ["BLIS_TIMING_OUT"] = "/data/timing.jsonl"

llm = LLM(
    model="Qwen/Qwen3-14B",
    tensor_parallel_size=4,        # adjust to your deployment
    enable_prefix_caching=False,   # keep off during data collection
    enforce_eager=False,           # keep CUDAGraphs ON — they reflect real production timing
)

# ── Warmup ────────────────────────────────────────────────────────────────────
# Run enough steps to saturate CUDAGraph compilation and JIT caches.
# Discard all records collected during warmup.
print("Warming up...")
llm.generate(["warmup prompt"] * 16, SamplingParams(max_tokens=10), use_tqdm=False)
open("/data/timing.jsonl", "w").close()   # truncate — warmup records are noise
print("Warmup done. Collecting timing data...")

# ── Sweep 1: Prefill — vary total prefill tokens ──────────────────────────────
# One request per shape, so num_reqs=1 and sum_ctx=0.
# The first step of each generate() call is the prefill step.
for prompt_len in [64, 128, 256, 512, 1024, 2048, 4096, 8192]:
    llm.generate(
        [{"prompt_token_ids": [1] * prompt_len}],
        SamplingParams(max_tokens=1, ignore_eos=True),
        use_tqdm=False,
    )

# ── Sweep 2: Decode — vary (batch_size B, context_length K) ──────────────────
# Submit B requests with a K-token prompt and generate 1 token.
# The first step is a prefill step (sum_ctx=0); the SECOND step (if max_tokens≥2)
# is a decode step at sum_ctx ≈ B*K.
# To isolate decode, use max_tokens=2 and keep only records where sum_ctx > 0.
for B in [1, 2, 4, 8, 16, 32, 64]:
    for K in [128, 256, 512, 1024, 2048, 4096]:
        if B * K > llm.llm_engine.model_config.max_model_len:
            continue
        llm.generate(
            [{"prompt_token_ids": [1] * K}] * B,
            SamplingParams(max_tokens=2, ignore_eos=True),
            use_tqdm=False,
        )

# ── Sweep 3: Mixed — prefill + decode in the same step ───────────────────────
# Submit some requests mid-decode alongside a new arriving request.
# This is harder to control precisely via generate(); easier via continuous serving.
# For a first approximation, the prefill and decode sweeps above are sufficient
# to fit β₁–β₇. Mixed steps refine β coefficients further.

print("Data collection complete.")
print(f"Records written: {sum(1 for _ in open('/data/timing.jsonl'))}")
```

### Understanding the output

Each line in `timing.jsonl` is a JSON object:

```json
{"num_tokens": 512, "num_reqs": 1, "sum_ctx_tokens": 0, "step_time_us": 4823.1}
{"num_tokens": 1,   "num_reqs": 1, "sum_ctx_tokens": 512, "step_time_us": 312.4}
```

- `sum_ctx_tokens == 0` → prefill step (all tokens are new prompt tokens)
- `sum_ctx_tokens > 0` → decode step (`num_reqs` requests each generating one new token, with cumulative context of `sum_ctx_tokens`)

The first record above: a 512-token prompt processed in 4.8 ms.
The second: a single decode step reading back 512 tokens of KV cache in 0.3 ms.

---

### TP: no special handling needed

With TP=N, vLLM spawns N worker processes (one per GPU). The plugin runs in all N workers and patches all N copies of `GPUModelRunner`. The `is_global_first_rank()` guard ensures only rank 0 writes records.

**Why rank 0 timing is correct for TP:**

Inside `_model_forward`, every transformer sublayer (attention + FFN) ends with an NCCL All-Reduce over the TP group. An All-Reduce cannot complete on rank 0 until all N ranks have contributed their shard. So rank 0's `end` CUDA event only fires after all TP ranks have finished their portion of the forward pass. The measured `step_time_us` therefore includes TP communication cost (`β₄·T_tp` in a physics model). No extra `torch.distributed.barrier()` is needed.

### EP: same approach, one nuance

With expert parallelism, token dispatch and gather use NCCL All-to-All. The All-to-All on rank 0 waits for all expert ranks to process the tokens dispatched to them. So rank 0's timing again captures total step time including EP communication.

The one nuance: if experts are unevenly loaded (some experts receive more tokens), the step time on rank 0 reflects the slowest expert rank. This is the correct behavior to include in training data — a latency model should predict the actual wall-clock step time including load imbalance, not an idealized balanced estimate. Models with EPLB (Expert Load Balancing) enabled will produce lower, more stable measurements.

### Multi-node

For multi-node deployments, only the node containing rank 0 (global rank 0 across all PP/TP/DP dimensions) writes records. The `is_global_first_rank()` check handles this automatically. Other nodes discard their timing measurements.

---

## Frequently Asked Questions

**Q: Will this slow down inference?**

Each recorded step adds one `torch.cuda.synchronize()` call. This blocks the CPU until the GPU finishes the forward pass, which adds zero GPU latency but does serialize the CPU pipeline. During data collection (offline sweeps) this is acceptable. For production serving, uninstall the plugin (`pip uninstall blis-timing-plugin`) or set `BLIS_TIMING_OUT=/dev/null` and add an early return to `_timed_model_forward` when the file is `/dev/null`.

A lighter-weight production variant would use `start_event.elapsed_time(end_event)` asynchronously (without `synchronize()`) and flush records in a background thread, but for offline training data collection the simple synchronous approach is fine.

---

**Q: Does this work with `enforce_eager=True` (no CUDAGraphs)?**

Yes, and the measurements will still be correct. However, CUDAGraph replay is what production vLLM actually uses for decode steps. Training data collected with `enforce_eager=True` will reflect eager-mode kernel overhead (higher, more variable) rather than CUDAGraph replay overhead (lower, more stable). For training a latency model that predicts production behavior, keep `enforce_eager=False` (the default).

---

**Q: What about the v2 model runner (`VLLM_USE_V2_MODEL_RUNNER=1`)?**

The v2 runner (`vllm/v1/worker/gpu/model_runner.py`) does not have a `_model_forward` method. It is disabled by default (`VLLM_USE_V2_MODEL_RUNNER=0`). If you enable it, patch `execute_model` on the v2 `GPUModelRunner` class instead. The timing and TP logic is identical; only the class import path changes:

```python
# v2 runner patch target (only needed if VLLM_USE_V2_MODEL_RUNNER=1):
from vllm.v1.worker.gpu.model_runner import GPUModelRunner as V2GPUModelRunner
```

---

**Q: Do I need to do anything different per TP degree?**

The plugin code is identical for TP=1, TP=4, TP=8, etc. What changes is the **training data you collect**: the β coefficients in a physics latency model encode hardware-specific ratios (e.g. NVLink bandwidth vs HBM bandwidth for the TP term). These ratios are TP-degree-dependent. Collect a separate dataset for each TP configuration you intend to deploy against, and fit separate coefficient sets.

---

**Q: The output file doesn't exist / is empty.**

Check that:
1. `pip show blis-timing-plugin` confirms the package is installed in the same environment vLLM uses.
2. `python -c "from importlib.metadata import entry_points; print([e.name for e in entry_points(group='vllm.general_plugins')])"` shows `blis_timing`.
3. The worker process has write permission to the directory in `BLIS_TIMING_OUT`.
4. Warmup completed before the file was truncated.

---

**Q: I see duplicate records / records from multiple ranks.**

The `is_global_first_rank()` guard should prevent this. If you see duplicates, check whether `torch.distributed` failed to initialize (in which case all ranks return True from that function). You can add a defensive check:

```python
import torch.distributed as dist
if dist.is_initialized() and dist.get_rank() != 0:
    return _original_model_forward(...)
```

---

## Summary

| What | How |
|---|---|
| Hook point | `GPUModelRunner._model_forward` (explicitly designed for override) |
| Timing method | CUDA events (`start.elapsed_time(end) * 1000` → µs) |
| Loaded in workers | via `vllm.general_plugins` entry point in `pyproject.toml` |
| TP/EP correctness | NCCL collectives inside forward() act as implicit barriers; rank 0 timing captures full group latency |
| Rebuild required | **No** — pure Python, `pip install -e .` only |
| Remove | `pip uninstall blis-timing-plugin` |
| Control output path | `BLIS_TIMING_OUT=/your/path.jsonl` environment variable |

The complete plugin is ~30 lines of Python and one `pyproject.toml`. The sweep script drives any batch shape combination through `llm.generate()`. The output is a JSONL file of `(num_tokens, num_reqs, sum_ctx_tokens, step_time_us)` records ready for regression.
